tree/treeplayer/src/TTreeFormula.cxx

```diff
@@ -3412,8 +3412,24 @@ Int_t TTreeFormula::GetRealInstance(Int_t instance, Int_t codeindex) {
                }
                break;
             case -1: {
-                  local_index = 0;
-                  Int_t virt_accum = 0;
+	      // caches to keep track of where we are across function invocations
+	      // will avoid an O(N^2) scaling
+	      // reset them if instance=0 is seen
+              static Int_t local_index_cache = 0;
```

Comment:
Yes of course. Let us focus on the correctness for one thread for the moment. The question is then if we want to use static stuff for the cache or a member variable of the TTreeFormula. The latter would not have a penalty to access (probably).